signer/signer.go

```diff
@@ -0,0 +1,134 @@
+package signer
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	msgsigner "github.com/form3tech-oss/go-http-message-signatures"
+	"github.com/form3tech-oss/http-message-signing-proxy/config"
+	"github.com/form3tech-oss/http-message-signing-proxy/proxy"
+)
+
+const (
+	requestTargetHeaderKey = "(request-target)"
+	digestHeaderKey        = "digest"
+)
+
+type DataError = msgsigner.DataError
+
+type requestSigner struct {
+	*msgsigner.MessageSigner
+	headerConfig config.HeadersConfig
+}
+
+func NewRequestSigner(cfg config.SignerConfig) (proxy.RequestSigner, error) {
+	key, err := loadKey(cfg.KeyFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	signatureHashAlgo, err := getHashAlgo(cfg.SignatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	digestHashAlgo, err := getHashAlgo(cfg.SignatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	signer, err := msgsigner.NewRSASigner(key, signatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	msgSigner, err := msgsigner.NewMessageSigner(digestHashAlgo, signer, cfg.KeyId, msgsigner.Signature)
+	if err != nil {
+		return nil, err
+	}
+
+	return &requestSigner{
+		MessageSigner: msgSigner,
+		headerConfig:  cfg.Headers,
+	}, err
+}
+
+func (rs *requestSigner) SignRequest(req *http.Request) (*http.Request, error) {
+	headers, err := rs.getSignatureHeaders(req)
+	if err != nil {
+		return nil, err
+	}
+
+	signedReq, err := rs.MessageSigner.SignRequest(req, headers)
+	switch err.(type) {
+	case *msgsigner.DataError:
+		return nil, proxy.NewInvalidRequestError(err)
+	default:
+		return signedReq, err
+	}
+}
+
+func (rs *requestSigner) getSignatureHeaders(req *http.Request) ([]string, error) {
+	// Get the intersection of request's headers and config's signature headers
+	var headers []string
+	for _, header := range rs.headerConfig.SignatureHeaders {
+		if req.Header.Get(header) != "" {
+			headers = append(headers, header)
+		}
+	}
+	if len(headers) == 0 {
+		return nil, proxy.NewInvalidRequestError(fmt.Errorf("none of the signature headers found in the request, expected at least one"))
+	}
+
+	// Include 'digest' header for PUT, POST and PATCH requests only
+	if rs.headerConfig.IncludeDigest && shouldHaveBody(req) {
```

Comment:
this will fail the signing if the request doesn't actually have a body. I think `shouldHaveBody` should check that `req` has a non-nil/non-empty body

Reply:
@samphillips-f3 I will check if there is a way to do it without reading the body.
@mszczygiel-form3 Why don't we support that?

Reply:
I added a check for empty body in `shouldHaveBody` function. `req.Body != nil && req.Body != http.NoBody`
Now if the request is put/post/patch and the body is empty, the digest header won't be added

Reply:
I'm not sure how it would work now for `chunked` requests now, but I assume we would read all the chunks and add signature then send it to the destination as a single request? It would make it easy to break this service (by transferring chunks infinitely)

Reply:
chunked request means a request that is split to multiple requests?

Reply:
yes

Reply:
Hmm I can put that in the doc